pkg/splunk/enterprise/workerpool.go

```diff
@@ -0,0 +1,184 @@
+		S3Reponse:          s3Response,
+		availableDiskSpace: availableDiskSpace,
+		wg:                 new(sync.WaitGroup),
+	}
+	return w
+}
+
+// Run is the main work that workers do to install the apps.
+// Workers wait for the jobs on the jobs channel and pick them up as they come in
+func (w *AppInstallWorkerPool) Run(id int, jobs <-chan *enterpriseApi.AppDeploymentInfo) {
+	scopedLog := log.WithName("AppInstallWorkerPool.run()").WithValues("worker id", id)
+	for job := range jobs {
+		scopedLog.Info("Started install of app", "appName", job.AppName)
+
+		object := getRemoteObjectFromS3Response(job.AppName, w.S3Reponse)
+		localFile := getLocalAppFileName(w.DownloadPath, job.AppName, *object.Etag)
+		remoteFile := *object.Key
+
+		// download the app from remote storage
+		err := w.S3ClientMgr.DownloadApp(remoteFile, localFile, *object.Etag)
+		if err != nil {
+			scopedLog.Error(err, "unable to download app", "appName", job.AppName)
+			setAppInstallState(job, enterpriseApi.DownloadError)
+		} else {
+			// download is successfull, update the state
+			setAppInstallState(job, enterpriseApi.DownloadComplete)
+
+			//TODO: gaurav, need to fill in below code
+			// copy the app to the splunk pod
+
+			// untar it, if it is cluster scoped
+
+			// if everything is successful, delete the app from operator pod
+		}
+
+		scopedLog.Info("finished job")
+		w.mu.Lock()
+		w.availableDiskSpace = w.availableDiskSpace + job.Size
+		scopedLog.Info("availableDiskSpace after freeing up space", "availableDiskSpace", w.availableDiskSpace)
+		w.mu.Unlock()
+
+		w.wg.Done()
+	}
+}
+
+// isAppAlreadyDownloaded checks if the app is already downloaded on the operator pod by checking a couple of things -
+// 1. is the app bundle present on the operator pod &&
+// 2. if the size on disk matches the size reported from remote storage
+func (w *AppInstallWorkerPool) isAppAlreadyDownloaded(app *enterpriseApi.AppDeploymentInfo) bool {
+	scopedLog := log.WithName("isAppAlreadyDownloaded").WithValues("app name", app.AppName)
+	object := getRemoteObjectFromS3Response(app.AppName, w.S3Reponse)
+	localAppFileName := getLocalAppFileName(w.DownloadPath, app.AppName, *object.Etag)
+
+	var fileInfo fs.FileInfo
+	var err error
+	// 1. First check if the app is present on operator pod
+	if fileInfo, err = os.Stat(localAppFileName); os.IsNotExist(err) {
+		scopedLog.Info("App not present on operator pod")
+		return false
+	}
+
+	// 2. Now, since the app is present, check the size
+	localSize := fileInfo.Size()
+	remoteSize := *object.Size
+	if localSize != remoteSize {
+		err = fmt.Errorf("size of app on operator pod does not match size on remote storage. localSize: %d, remoteSize:%d", localSize, remoteSize)
+		scopedLog.Error(err, "need to re-download app")
+		return false
+	}
+
+	// If we reached here, this means app was downloaded successfuly to operator pod,
+	// mark this as DownloadComplete
+	setAppInstallState(app, enterpriseApi.DownloadComplete)
```

Comment:
Nit: To make the semantics clean, the caller of this API can call `setAppInstallState`, when true

Reply:
Yeah I think either way its fine  TBH :)

Reply:
`isAppAlreadyDownloaded` sounds like `true/false` return kind :-). That is why, either we better rename it to something more logical OR, handle it at the calling place :-)